From 0b511d8cb3e5cac59eb6ad3d5a0c0bfe26c90b0f Mon Sep 17 00:00:00 2001
From: Tony Crisci <tony@dubstepdish.com>
Date: Thu, 1 May 2014 23:26:52 -0400
Subject: [PATCH] Implement the ipc 'binding' event

The binding event will be triggered when a binding is run as a result of
some a user action. The binding event has the following properties:

change: (str) Currently this will only be "run" but may be expanded in
the future. Included for consistency with other events.

binding: (map) the serialized binding

The "binding" member will have these properties:

input_type: (str) either "keyboard" or "mouse"

input_code: (int) the xcb keycode of the keyboard binding if it was
provided or the mouse button if it is a mouse binding.

symbol: (str) the string representation of the input code

command: (str) the bound command

mods: (list of str) a list of the modifiers that were pressed as string
symbols

fixes #1210
---
 include/bindings.h                  |  5 +++
 include/i3/ipc.h                    |  3 ++
 include/ipc.h                       |  5 +++
 src/bindings.c                      | 16 ++++++++
 src/ipc.c                           | 82 +++++++++++++++++++++++++++++++++++++
 src/key_press.c                     |  7 +---
 testcases/t/226-ipc-binding-event.t | 53 ++++++++++++++++++++++++
 7 files changed, 165 insertions(+), 6 deletions(-)
 create mode 100644 testcases/t/226-ipc-binding-event.t

diff --git a/include/bindings.h b/include/bindings.h
index d3d3aa8..2cf576f 100644
--- a/include/bindings.h
+++ b/include/bindings.h
@@ -57,3 +57,8 @@ void switch_mode(const char *new_mode);
  *
  */
 void check_for_duplicate_bindings(struct context *context);
+
+/**
+ * Runs the given binding and returns a CommandResult
+ */
+struct CommandResult *run_binding(Binding *bind);
diff --git a/include/i3/ipc.h b/include/i3/ipc.h
index 94a3990..d5f6ab4 100644
--- a/include/i3/ipc.h
+++ b/include/i3/ipc.h
@@ -100,3 +100,6 @@ typedef struct i3_ipc_header {
 
 /** Bar config update will be triggered to update the bar config */
 #define I3_IPC_EVENT_BARCONFIG_UPDATE           (I3_IPC_EVENT_MASK | 4)
+
+/** The binding event will be triggered when bindings run */
+#define I3_IPC_EVENT_BINDING                    (I3_IPC_EVENT_MASK | 5)
diff --git a/include/ipc.h b/include/ipc.h
index 418b040..a184503 100644
--- a/include/ipc.h
+++ b/include/ipc.h
@@ -99,3 +99,8 @@ void ipc_send_window_event(const char *property, Con *con);
  * For the barconfig update events, we send the serialized barconfig.
  */
 void ipc_send_barconfig_update_event(Barconfig *barconfig);
+
+/**
+ * For the binding events, we send the serialized binding struct.
+ */
+void ipc_send_binding_event(const char *event_type, Binding *binding);
diff --git a/src/bindings.c b/src/bindings.c
index 3d3dbd9..14fedb4 100644
--- a/src/bindings.c
+++ b/src/bindings.c
@@ -326,3 +326,19 @@ void check_for_duplicate_bindings(struct context *context) {
         }
     }
 }
+
+/*
+ * Runs the given binding and returns a CommandResult
+ */
+struct CommandResult *run_binding(Binding *bind) {
+    char *command_copy = sstrdup(bind->command);
+    struct CommandResult *command_output = parse_command(command_copy);
+    free(command_copy);
+
+    if (command_output->needs_tree_render)
+        tree_render();
+
+    ipc_send_binding_event("run", bind);
+
+    return command_output;
+}
diff --git a/src/ipc.c b/src/ipc.c
index 66c63f7..59c9e33 100644
--- a/src/ipc.c
+++ b/src/ipc.c
@@ -147,6 +147,57 @@ static void dump_rect(yajl_gen gen, const char *name, Rect r) {
     y(map_close);
 }
 
+static void dump_binding(yajl_gen gen, Binding *bind) {
+    y(map_open);
+    ystr("input_code");
+    y(integer, bind->keycode);
+
+    ystr("input_type");
+    ystr((const char*)(bind->input_type == B_KEYBOARD ? "keyboard" : "mouse"));
+
+    ystr("symbol");
+    ystr(bind->symbol);
+
+    ystr("command");
+    ystr(bind->command);
+
+    ystr("mods");
+    y(array_open);
+    for (int i = 0; i < 8; i++) {
+        if (bind->mods & (1 << i)) {
+            switch (1 << i) {
+                case XCB_MOD_MASK_SHIFT:
+                    ystr("shift");
+                    break;
+                case XCB_MOD_MASK_LOCK:
+                    ystr("lock");
+                    break;
+                case XCB_MOD_MASK_CONTROL:
+                    ystr("ctrl");
+                    break;
+                case XCB_MOD_MASK_1:
+                    ystr("Mod1");
+                    break;
+                case XCB_MOD_MASK_2:
+                    ystr("Mod2");
+                    break;
+                case XCB_MOD_MASK_3:
+                    ystr("Mod3");
+                    break;
+                case XCB_MOD_MASK_4:
+                    ystr("Mod4");
+                    break;
+                case XCB_MOD_MASK_5:
+                    ystr("Mod5");
+                    break;
+            }
+        }
+    }
+    y(array_close);
+
+    y(map_close);
+}
+
 void dump_node(yajl_gen gen, struct Con *con, bool inplace_restart) {
     y(map_open);
     ystr("id");
@@ -1114,6 +1165,37 @@ void ipc_send_barconfig_update_event(Barconfig *barconfig) {
     y(get_buf, &payload, &length);
 
     ipc_send_event("barconfig_update", I3_IPC_EVENT_BARCONFIG_UPDATE, (const char *)payload);
+
+    y(free);
+    setlocale(LC_NUMERIC, "");
+}
+
+/*
+ * For the binding events, we send the serialized binding struct.
+ */
+void ipc_send_binding_event(const char *event_type, Binding *bind) {
+    DLOG("Issue IPC binding %s event (sym = %s, code = %d)\n", event_type, bind->symbol, bind->keycode);
+
+    setlocale(LC_NUMERIC, "C");
+
+    yajl_gen gen = ygenalloc();
+
+    y(map_open);
+
+    ystr("change");
+    ystr(event_type);
+
+    ystr("binding");
+    dump_binding(gen, bind);
+
+    y(map_close);
+
+    const unsigned char *payload;
+    ylength length;
+    y(get_buf, &payload, &length);
+
+    ipc_send_event("binding", I3_IPC_EVENT_BINDING, (const char *)payload);
+
     y(free);
     setlocale(LC_NUMERIC, "");
 }
diff --git a/src/key_press.c b/src/key_press.c
index 68e2fca..aa71925 100644
--- a/src/key_press.c
+++ b/src/key_press.c
@@ -105,12 +105,7 @@ void handle_key_press(xcb_key_press_event_t *event) {
         }
     }
 
-    char *command_copy = sstrdup(bind->command);
-    struct CommandResult *command_output = parse_command(command_copy);
-    free(command_copy);
-
-    if (command_output->needs_tree_render)
-        tree_render();
+    struct CommandResult *command_output = run_binding(bind);
 
     /* We parse the JSON reply to figure out whether there was an error
      * ("success" being false in on of the returned dictionaries). */
diff --git a/testcases/t/226-ipc-binding-event.t b/testcases/t/226-ipc-binding-event.t
new file mode 100644
index 0000000..656d9c9
--- /dev/null
+++ b/testcases/t/226-ipc-binding-event.t
@@ -0,0 +1,53 @@
+#!perl
+# vim:ts=4:sw=4:expandtab
+#
+# Please read the following documents before working on tests:
+# • http://build.i3wm.org/docs/testsuite.html
+#   (or docs/testsuite)
+#
+# • http://build.i3wm.org/docs/lib-i3test.html
+#   (alternatively: perldoc ./testcases/lib/i3test.pm)
+#
+# • http://build.i3wm.org/docs/ipc.html
+#   (or docs/ipc)
+#
+# • http://onyxneon.com/books/modern_perl/modern_perl_a4.pdf
+#   (unless you are already familiar with Perl)
+#
+# Test that the binding event works properly
+# Ticket: #1210
+# Bug still in: 4.7.2-148-g54a6404
+use i3test i3_autostart => 0;
+use X11::Protocol;
+
+my $config = <<EOT;
+# i3 config file (v4)
+font -misc-fixed-medium-r-normal--13-120-75-75-C-70-iso10646-1
+
+bindcode 214 exec nop
+EOT
+
+my $pid = launch_with_config($config);
+
+my $i3 = i3(get_socket_path());
+$i3->connect->recv;
+
+my $xprotocol = X11::Protocol->new;
+$xprotocol->init_extension('XTEST') or BAIL_OUT "XTEST extension not available";
+
+my $cv = AE::cv;
+my $timer = AE::timer 0.5, 0, sub { $cv->send(0); };
+
+$i3->subscribe({
+        binding => sub {
+            $cv->send('got binding event');
+        }
+    })->recv;
+
+$xprotocol->XTestFakeInput( name => 'KeyPress', detail => 214 );
+
+ok($cv->recv, 'got binding event');
+
+does_i3_live;
+
+done_testing;
-- 
1.9.2

